text/0000-get-collection-keys.md

```diff
@@ -0,0 +1,121 @@
+- Feature Name: get_collection_keys
+- Start Date: 2015-06-24
+- RFC PR: https://github.com/rust-lang/rust/pull/26531
+- Rust Issue:
+
+# Summary
+
+Get the original keys back from collections.
+
+# Motivation
+
+Removing a key from a HashMap or an item from a HashSet will drop the item.
+This is usually acceptable for small clonable objects like numbers and strings,
+but they are not always clonable, and even if they are it might be an expensive
+operation.
```

Comment:
This needs to be massively fleshed out. I would argue that there are two major motivators:

* Sets 
* Entries

Entries would primarily like to be able to reflect on the key to complete the construction of the value. Potentially also to decide whether to *really* do the insertion, though I don't expect this much or at all.

Sets on the other hand would like to be able to talk about the key *at all*. In particular you might use a Set as a cache by indexing on only a small fragment of the key (e.g. an integer id). For technical reasons it may be infeasible or just ineffecient to split the key into a proper (key, value) pair.

The rest of Maps are for the most part, I think, just a nice-to-have. Largely the extended Map API is just needed to implement the new Set API on top of (since Sets are generally trivial Map wrappers).